### ISSUE TITLE
In-kernel forwarding doesn't work mr9.5.1.2, CentOS 8 (4.18.0-305.10.2.el8_4.x86_64)

### ISSUE THREAD

**Issue by abalashov** [sipwise/rtpengine#1329]
With RTPEngine in between endpoints `A` and `B`, after a few seconds of "learning", `A -> RTPEngine` doesn't make it out to the `B` side, and `B -> RTPEngine` doesn't make it out the `A` side. 

I am using two interfaces (`-i external/x.x.x.x -i internal/y.y.y.y`) and setting ingress and egress parameters for them correctly--a packet capture confirms the SDP substitution is happening correctly. The problem presents equally well with one interface, however.

I initially deployed RTPEngine on CentOS 8 in order to link it against OpenSSL 1.1.1 to solve what appear to be recent browser endpoint changes around RFC7714/AEAD, which were solved by doing this. At least, I think it's browser endpoint changes, because JsSIP endpoints <= mr8.5 suddenly stopped negotiating DTLS successfully.

However, this problem has nothing to do with encryption, insofar as it occurs with plain old RTP streams as well.

```
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: NOTICE: [aj9heq3su9rv0tk6chvg]: [core] Creating new call
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: WARNING: [aj9heq3su9rv0tk6chvg]: [core] Ignoring updated remote endpoint  as the local socket is IPv4
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg]: [control] Replying to 'offer' from 10.150.21.15:51747 (elapsed time 0.001620 sec)
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg]: [control] Received command 'offer' from 10.150.21.15:51747
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: WARNING: [aj9heq3su9rv0tk6chvg]: [core] Ignoring updated remote endpoint  as the local socket is IPv4
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg]: [control] Replying to 'offer' from 10.150.21.15:51747 (elapsed time 0.000230 sec)
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg]: [control] Received command 'answer' from 10.150.21.15:48376
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg]: [control] Replying to 'answer' from 10.150.21.15:48376 (elapsed time 0.000128 sec)
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: ERR: [aj9heq3su9rv0tk6chvg port  8088]: [core] SRTP output wanted, but no crypto suite was negotiated
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [ice] ICE negotiated: peer for component 1 is 76.125.16.160:49714
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [ice] ICE negotiated: local interface 66.42.86.248
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [crypto] DTLS: Peer certificate accepted
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [crypto] DTLS-SRTP successfully negotiated using AEAD_AES_256_GCM
Aug 11 04:04:58 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [crypto] DTLS-SRTP successfully negotiated using AEAD_AES_256_GCM
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [core] Confirmed peer address as 76.125.16.160:49714
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8088]: [core] Confirmed peer address as 10.150.21.10:22766
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8088]: [core] Kernelizing media stream: 10.150.21.10:22766
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [core] Kernelizing media stream: 76.125.16.160:49714
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8089]: [core] Confirmed peer address as 10.150.21.10:22767
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [core] Confirmed peer address as 76.125.16.160:49714
Aug 11 04:05:02 rtpenginetest rtpengine[26773]: INFO: [aj9heq3su9rv0tk6chvg port  8082]: [core] Kernelizing media stream: 76.125.16.160:49714
```

It's when the media stream becomes "kernelized" that things go off the rails. Until then, bidirectional media is passed.

Suspicion falls upon this message:

```
Ignoring updated remote endpoint  as the local socket is IPv4
```

Looking at the source, there's a placeholder for a substitution token after the word `endpoint`, so there should be a value there. I'm not sure why there's not, or where to look.

I did look into `ip_forward` and `rp_filter` kernel options, though changing them has never regulated RTPEngine behaviour in the past for me, and this time is no exception.

Thank you kindly for any time and attention you are able to give to help. Anything you can spare is deeply appreciated.

**Comment by rfuchs** [comment]
Inspect the contents of /proc/rtpengine/0/list after a stream has been pushed to the kernel

**Comment by abalashov** [comment]
Everything looks normal, packet and octet counters are incrementing on both streams.

But on the wire, the forwarding isn’t happening. Packets are coming in from both endpoints and not leaving to the other endpoint.

—

**Comment by rfuchs** [comment]
Perhaps kernel support for AEAD is missing or broken. Inspect /proc/crypto for an entry that looks like this:
```
name         : gcm(aes)
driver       : generic-gcm-aesni
module       : aesni_intel
priority     : 400
refcnt       : 1
selftest     : passed
internal     : no
type         : aead
async        : yes
blocksize    : 1
ivsize       : 12
maxauthsize  : 16
geniv        : <none>
```
(Drive and module might be different)

**Comment by abalashov** [comment]
Hi,

This is what I've got for `gcm(aes)`:

```
name         : gcm(aes)
driver       : gcm_base(ctr(aes-aesni),ghash-generic)
module       : kernel
priority     : 200
refcnt       : 1
selftest     : passed
internal     : no
type         : aead
async        : no
blocksize    : 1
ivsize       : 12
maxauthsize  : 16
geniv        : <none>
```

**Comment by rfuchs** [comment]
LGTM. Not sure what could be going wrong then. The only significant change in the kernel module between 8.5 and 9.5 was the addition of AEAD support. Some strange firewall rules or routing setup maybe? But if it works from userspace it should work from kernel as well.

**Comment by abalashov** [comment]
Thanks, I’ll see what else I can figure out.

Can you confirm that the browsers started mandating AEAD? I had not heard anything about this, but DTLS stopped working - over time - for JsSIP endpoints—until we upgraded to precisely the RTPEngine version that supports AEAD.

—

**Comment by rfuchs** [comment]
> Can you confirm that the browsers started mandating AEAD? I had not heard anything about this, but DTLS stopped working - over time - for JsSIP endpoints—until we upgraded to precisely the RTPEngine version that supports AEAD.

Not that I've heard myself, but I'm not following the most recent browser developments that closely.

**Comment by abalashov** [comment]
It was something inexplicable with the host. I tried it on a CentOS 8 AWS instance and it worked fine. I'm going to close this item since it doesn't seem reproducible outside the particular host in question. Thanks for your help!

**Comment by themsley-voiceflex** [comment]
We also experienced a similar problem but far less frequently - every couple of days or so. Noticed a mismatch in our iptables rules vs the port range specified in rtpengine.conf and have changed them to match and have not had a repeat of the problem since (5 days so far). iptables allows udp ports 10000:54000, rtpengine.conf said 10000:60000. I suspect that we got the problem when it picked ephemeral ports > 54000 & less than 60k.
